File: ruby/lib/steam/packets/s2a_info2_packet.rb
# This code is free software; you can redistribute it and/or modify it under
# the terms of the new BSD License.
#
# Copyright (c) 2008-2011, Sebastian Staudt

require 'steam/packets/s2a_info_base_packet'

# This class represents a S2A_INFO_DETAILED response packet sent by a Source or
# GoldSrc server
#
# Out-of-date (before 10/24/2008) GoldSrc servers use an older format (see
# {S2A_INFO_DETAILED_Packet}).
#
# @author Sebastian Staudt
# @see GameServer#update_server_info
class S2A_INFO2_Packet

  include S2A_INFO_BasePacket

  # Creates a new S2A_INFO2 response object based on the given data
  #
  # @param [String] data The raw packet data replied from the server
  # @see S2A_INFO_BasePacket#generate_info_hash
  def initialize(data)
    super S2A_INFO2_HEADER, data

    @protocol_version = @content_data.byte
    @server_name = @content_data.cstring
    @map_name = @content_data.cstring
    @game_directory = @content_data.cstring
    @game_description = @content_data.cstring
    @app_id = @content_data.short
    @number_of_players = @content_data.byte
    @max_players = @content_data.byte
    @number_of_bots = @content_data.byte
    @dedicated = @content_data.byte.chr
    @operating_system = @content_data.byte.chr
    @password_needed = @content_data.byte == 1
    @secure = @content_data.byte == 1
    @game_version = @content_data.cstring

    if @content_data.remaining > 0
      extra_data_flag = @content_data.byte

      @server_port = @content_data.short unless extra_data_flag & 0x80 == 0

      unless extra_data_flag & 0x10 == 0
        @server_id =  @content_data.long | (@content_data.long << 32)
      end

      unless extra_data_flag & 0x40 == 0
        @tv_port = @content_data.short
        @tv_name = @content_data.cstring
      end

      @server_tags = @content_data.cstring unless extra_data_flag & 0x20 == 0
    end

    generate_info_hash
  end

end

File: php/lib/steam/packets/S2A_INFO2_Packet.php
<?php
/**
 * This code is free software; you can redistribute it and/or modify it under
 * the terms of the new BSD License.
 *
 * @author     Sebastian Staudt
 * @license    http://www.opensource.org/licenses/bsd-license.php New BSD License
 * @package    Steam Condenser (PHP)
 * @subpackage Packets
 */

require_once STEAM_CONDENSER_PATH . 'steam/packets/S2A_INFO_BasePacket.php';

/**
 * @package Steam Condenser (PHP)
 * @subpackage Packets
 */
class S2A_INFO2_Packet extends S2A_INFO_BasePacket
{
	/**
	 *
	 */
	public function __construct($data)
	{
		parent::__construct(SteamPacket::S2A_INFO2_HEADER, $data);

		$this->networkVersion = $this->contentData->getByte();
		$this->serverName = $this->contentData->getString();
		$this->mapName = $this->contentData->getString();
		$this->gameDir = $this->contentData->getString();
		$this->gameDesc = $this->contentData->getString();
		$this->appId = $this->contentData->getShort();
		$this->playerNumber = $this->contentData->getByte();
		$this->maxPlayers = $this->contentData->getByte();
		$this->botNumber = $this->contentData->getByte();
		$this->dedicated = chr($this->contentData->getByte());
		$this->operatingSystem = chr($this->contentData->getByte());
		$this->passwordProtected = $this->contentData->getByte() == 1;
		$this->secureServer = $this->contentData->getByte() == 1;
		$this->gameVersion = $this->contentData->getString();

		if($this->contentData->remaining() > 0)
		{
			$extraDataFlag = $this->contentData->getByte();
			 
			if($extraDataFlag & 0x80)
			{
				$this->serverPort = $this->contentData->getShort();
			}

            if($extraDataFlag & 0x10) {
                $this->serverId = $this->contentData->getLong() | ($this->contentData->getLong() << 32);
            }

			if($extraDataFlag & 0x40)
			{
				$this->tvPort = $this->contentData->getShort();
				$this->tvName = $this->contentData->getString();
			}
			 
			if($extraDataFlag & 0x20)
			{
				$this->serverTags = $this->contentData->getString();
			}
		}
	}
}
?>
